### PR TITLE
implemented SHSUB8 and SHSUB16.

### PR DESCRIPTION
**Pull request by FernandoS27**

### CODE REVIEW

File: src/backend_x64/emit_x64.cpp [hunk 1]
```diff
@@ -1853,6 +1853,42 @@ void EmitX64::EmitPackedHalvingSubU8(IR::Block&, IR::Inst* inst) {
...
+    // We must now perform a partitioned subtraction.
+    // We can do this because minuend contains 7 bit fields.
+    // We use the extra bit in minuend as a bit to borrow from; we set this bit.
+    // We invert this bit at the end as this tells us if that bit was borrowed from.
+    // Finaly, we apply back the carry  to guarantee.
+    code->or(minuend, 0x80808080);
+    code->sub(minuend, subtrahend);
+    code->xor(minuend, 0x80808080);
+    code->xor(minuend, carry);
+    // minuend now contains the desired result.
```
Comment (lioncash) [inline src/backend_x64/emit_x64.cpp]: Comments like these don't particularly elucidate anything that wasn't already known. It's expected that at the end of a routine, everything that needed to be done was done correctly.

File: src/backend_x64/emit_x64.cpp [hunk 2]
```diff
@@ -1853,6 +1853,42 @@ void EmitX64::EmitPackedHalvingSubU8(IR::Block&, IR::Inst* inst) {
...
+void EmitX64::EmitPackedHalvingSubS8(IR::Block&, IR::Inst* inst) {
+    IR::Value a = inst->GetArg(0);
+    IR::Value b = inst->GetArg(1);
+
+    Xbyak::Reg32 minuend = reg_alloc.UseDefGpr(a, inst).cvt32();
+    Xbyak::Reg32 subtrahend = reg_alloc.UseScratchGpr(b).cvt32();
+
+    Xbyak::Reg32 carry = reg_alloc.ScratchGpr().cvt32();
+
+    // This relies on the equality x-y == (x^y) - (((x^y)&y) << 1).
```
Comment (lioncash) [inline src/backend_x64/emit_x64.cpp]: These arithmetic expressions in comments should follow the same coding style (with regards to spacing) as regular code for consistency.

File: src/backend_x64/emit_x64.cpp (hunk 1 above)
Comment (FernandoS27) [inline src/backend_x64/emit_x64.cpp]: I agree, but then again it's the same as current implementation of SubU8.

File: src/backend_x64/emit_x64.cpp [hunk 3]
```diff
@@ -1875,11 +1909,45 @@ void EmitX64::EmitPackedHalvingSubU16(IR::Block&, IR::Inst* inst) {
...
+    // At this point,
+    // minuend := (a^b) >> 1
+    // subtrahend := (a^b) & b
+    // carry := (a^b) & 0x80008000
 
+    // We must now perform a partitioned subtraction.
+    // We can do this because minuend contains 7 bit fields.
+    // We use the extra bit in minuend as a bit to borrow from; we set this bit.
+    // We invert this bit at the end as this tells us if that bit was borrowed from.
+    // Finaly, we apply back the carry to guarantee.
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Missed changing this.

File: src/backend_x64/emit_x64.cpp [hunk 4]
```diff
@@ -1875,11 +1909,45 @@ void EmitX64::EmitPackedHalvingSubU16(IR::Block&, IR::Inst* inst) {
     // We must now perform a partitioned subtraction.
     // We can do this because minuend contains 15 bit fields.
     // We use the extra bit in minuend as a bit to borrow from; we set this bit.
-    // We invert this bit at the end as this tells us if that bit was borrowed from.
+    // We then sign extend the result into this bit.
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Don't change this.

**Pull request merged**